core/txn/pool/gatherer.go

```diff
@@ -0,0 +1,76 @@
+package pool
+
+import (
+	"context"
+	"sync"
+
+	"go.dedis.ch/dela/core/txn"
+)
+
+// Gatherer is a common tool to the pool implementations that helps to implement
+// the gathering process.
+type Gatherer interface {
+	// Wait waits for a notification with sufficient transactions to return the
+	// array, or nil if the context ends.
+	Wait(ctx context.Context, cfg Config) []txn.Transaction
+
+	// Notify triggers any wait that fits the given lenght. The function is only
+	// called when necessary.
+	Notify(length int, fn func() []txn.Transaction)
```

Comment:
are you using a callback for performance reasons? If so I would explain why in the comment. I guess it's a good reason, but it make the code painful to read and understand.

Reply:
Yes, I didn't want to make an array every time a transaction is added. But actually I had to change the gatherer in the next PR because of the locks (txs could be ignored) and the notify will disappear anyway.